test/test_databag.py

```diff
@@ -0,0 +1,527 @@
+        return MyDatabag
+
+    @property
+    def encoder(self) -> Callable[..., Any] | None:
+        return nested_encode
+
+    @property
+    def decoder(self) -> Callable[..., Any] | None:
+        return nested_decode
+
+
+@dataclasses.dataclass
+class MyDataclassDatabag(ops.DatabagBase):
+    foo: str
+    baz: list[str] = dataclasses.field(default_factory=list)
+    bar: int = dataclasses.field(default=0)
+    quux: Nested = dataclasses.field(default_factory=Nested)
+
+    def __post_init__(self):
+        if not isinstance(self.foo, str):
+            raise ValueError('foo must be a string')
+        if not isinstance(self.baz, list):
+            raise ValueError('baz must be a list')
+        if not all(isinstance(i, str) for i in self.baz):
+            raise ValueError('baz must be a list of strings')
+        if not isinstance(self.bar, int) or self.bar < 0:
+            raise ValueError('bar must be a zero or positive int')
+
+    def __setattr__(self, key: str, value: Any):
+        if key == 'bar' and (not isinstance(value, int) or value < 0):
+            raise ValueError('bar must be a zero or positive int')
+        super().__setattr__(key, value)
+
+
+class MyDataclassCharm(BaseTestCharm):
+    @property
+    def databag_class(self) -> type[ops.DatabagBase]:
+        return MyDataclassDatabag
+
+
+_test_classes: list[type[ops.CharmBase]] = [MyCharm, MyDataclassCharm]
+
+if pydantic:
+
+    @pydantic.dataclasses.dataclass
+    class MyPydanticDataclassDatabag(ops.DatabagBase):
+        foo: str
+        baz: list[str] = pydantic.Field(default_factory=list)
+        bar: int = pydantic.Field(default=0)
+        quux: Nested = pydantic.Field(default_factory=Nested)
+
+        def __post_init__(self):
+            if not isinstance(self.foo, str):
+                raise ValueError('foo must be a string')
+            if not isinstance(self.baz, list):
+                raise ValueError('baz must be a list')
+            if not all(isinstance(i, str) for i in self.baz):
+                raise ValueError('baz must be a list of strings')
+            if not isinstance(self.bar, int) or self.bar < 0:
+                raise ValueError('bar must be a zero or positive int')
+
+        # TODO: This should be done pydantic-style.
+        def __setattr__(self, key: str, value: Any):
+            if key == 'bar' and (not isinstance(value, int) or value < 0):
+                raise ValueError('bar must be a zero or positive int')
+            super().__setattr__(key, value)
+
+    class MyPydanticDataclassCharm(BaseTestCharm):
+        @property
+        def databag_class(self) -> type[ops.DatabagBase]:
+            return MyPydanticDataclassDatabag
+
+    class MyPydanticDatabag(pydantic.BaseModel, ops.DatabagBase):
+        foo: str
+        baz: list[str] = pydantic.Field(default_factory=list)
+        bar: int = pydantic.Field(default=0, ge=0)
+        quux: Nested = pydantic.Field(default_factory=Nested)
+
+        class Config:
```

Comment:
Consider:

```
model_config = ConfigDict(
            # tolerate additional keys in databag
            extra="ignore",
            # Allow instantiating this class by field name (instead of forcing alias).
            populate_by_name=True,
        )
```